dex/networks/btc/descriptors.go

```diff
@@ -0,0 +1,338 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package btc
+
+import (
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/btcsuite/btcd/btcec/v2"
+	"github.com/btcsuite/btcd/btcutil"
+	"github.com/btcsuite/btcd/btcutil/hdkeychain"
+)
+
+var (
+	// ErrMalformedDescriptor is returned when a provided descriptor string does
+	// not match the expected format for a descriptor.
+	ErrMalformedDescriptor = errors.New("malformed descriptor")
+)
+
+var (
+	hardened uint32 = hdkeychain.HardenedKeyStart
+
+	// {function}([{fingerprint}/{path}]{keyScriptOrNested})#{checksum}
+	descRE = regexp.MustCompile(`([[:alnum:]]+)` + `\((?:\[([[:xdigit:]]{8})/?([\d'h/]*)\])?(\S*)\)` +
+		`(?:#([[:alnum:]]{8}))?`) // https://regex101.com/r/75FEc4/1
+
+	// {xkey}{path}{range} e.g. {tpubDCD...}{/1'/2}{/*}
+	extKeyRE = regexp.MustCompile(`([[:alnum:]]+)((?:/\d+['h]?)*)(/\*['h]?)?`) // https://regex101.com/r/HLNzFF/1
+)
+
+// KeyOrigin describes the optional part of a KEY expression that contains key
+// origin information in side functions like pkh(KEY). For example, in the
+// descriptor wpkh([b940190e/84'/1'/0'/0/0]0300034...) the key origin is
+// [b940190e/84'/1'/0'/0/0], where the first part must be 8 hexadecimal
+// characters for the fingerprint of the master key, and there are zero or more
+// derivation steps to reach the key. See
+// https://github.com/bitcoin/bitcoin/blob/master/doc/descriptors.md#key-origin-identification
+// Note for Fingerprint: "software must be willing to deal with collisions".
+type KeyOrigin struct {
+	// Fingerprint is "Exactly 8 hex characters for the fingerprint of the key
+	// where the derivation starts".
+	Fingerprint string
```

Comment:
Notably, 
> //!< First 32 bits of the Hash160 of the public key at the root of the path

https://github.com/bitcoin/bitcoin/blob/01e9e2d1ca7fc08d65663b398c71ecec6a01f686/src/script/keyorigin.h#L13

So one could conceivably devise a scheme where we don't do path examination of the `listdescriptors true` result, but instead just match a key to the fingerprint. I only mention this because when I run `listdescriptors true` on the beta wallet, I notice that the same private key is listed for all functions.